olp-cpp-sdk-dataservice-read/src/VersionedLayerClientImpl.cpp

```diff
@@ -164,6 +167,164 @@ client::CancellableFuture<DataResponse> VersionedLayerClientImpl::GetData(
+            ExtendedApiResponse<PartitionDataHandleResult, client::ApiError,
+                                client::NetworkStatistics>;
+        repository::PartitionsRepository repository(
+            catalog, layer_id, *shared_settings, lookup_client);
+
+        auto query = [=](std::vector<std::string> root,
+                         client::CancellationContext inner_context) mutable
+            -> PartitionDataHandleResponse {
+          auto request = PartitionsRequest()
+                             .WithPartitionIds(std::move(root))
+                             .WithBillingTag(billing_tag);
+          auto response = repository.GetVersionedPartitions(request, version,
+                                                            inner_context);
+          if (!response.IsSuccessful()) {
+            OLP_SDK_LOG_WARNING_F(kLogTag,
+                                  "PrefetchPartitions: getting versioned "
+                                  "partitions failed, key=%s",
+                                  request.CreateKey(layer_id).c_str());
+            return response.GetError();
+          }
+          PartitionDataHandleResult result;
+          std::transform(std::begin(response.GetResult().GetPartitions()),
+                         std::end(response.GetResult().GetPartitions()),
+                         std::inserter(result, result.end()),
+                         [&](const model::Partition& partition) {
+                           return std::make_pair(partition.GetPartition(),
+                                                 partition.GetDataHandle());
+                         });
+          return PartitionDataHandleResponse(result);
+        };
+
+        auto download = [=](std::string data_handle,
+                            client::CancellationContext inner_context) mutable {
+          if (data_handle.empty()) {
+            return BlobApi::DataResponse(
+                client::ApiError(client::ErrorCode::NotFound, "Not found"));
+          }
+          repository::DataCacheRepository data_cache_repository(
+              catalog, shared_settings->cache);
+          if (data_cache_repository.IsCached(layer_id, data_handle)) {
+            return BlobApi::DataResponse(nullptr);
+          }
+
+          repository::DataRepository repository(catalog, *shared_settings,
+                                                lookup_client);
+          // Fetch from online
+          return repository.GetVersionedData(layer_id,
+                                             DataRequest()
+                                                 .WithDataHandle(data_handle)
+                                                 .WithBillingTag(billing_tag),
+                                             version, inner_context);
+        };
+
+        // split prefetch ids to blocks by 100 items
+        auto size_left = request.GetPartitionIds().size();
+        std::vector<std::vector<std::string>> roots;
```

Comment:
I don't like this. Why don't you split and trigger download directly? This way you do not need any vector in vector which looks awful.